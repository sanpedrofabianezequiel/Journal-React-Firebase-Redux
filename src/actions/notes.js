import Swal from "sweetalert2";
import { db } from "../firabase/firebase-config"
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


export const startNewNote = ()=>{
    return async (dispatch,getState) =>{
        //If the user is aunthenticated, we will set that ID how a new "namr or id name Collection"
        const {uid} = getState().auth; //is the samen if we use only destructuration
        const newNote = {
            title:'',
            body:'',
            date:new Date().getTime()
        }

        try {
            
            const docRef= await db.collection(`${uid}/journal/notes`).add(newNote);
           // console.log(docRef);
           dispatch(activeNote(docRef.id,newNote));//Envia el OBJ Action y llena el dispatch
           dispatch(addNewNote(docRef.id,newNote))
        } catch (error) {
            console.log(error);
        }
    }
}

export const addNewNote = (id,note)=>({
    type:types.notesAddNew,
    payload:{
        id,
        ...note
    }
})

export const activeNote = (id,note) =>({
    type: types.notesActive,
    payload:{
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) =>{
    return async (dispatch) =>{
        const notes = await loadNotes(uid);
        dispatch(setNote(notes));
    }
}
export const setNote = (notes)=>({
    type:types.notesLoad,
    payload: notes
})

export const startSaveNote = (note)=>{ //Actualizo x el id de usuario y id de note
    return async(dispatch,getState) =>{

        const {uid} = getState().auth; //Obtendo el ID del usuario o collecion q voy a guardar en firebas

        if(!note.url){
            delete note.url //si no esta direcetamente no la descompongo
        }

        const noteToFirestore = {...note}

        delete noteToFirestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
        dispatch(refresNote(note.id, noteToFirestore));//Podriamos enviar el note tambien
        Swal.fire('Saved', note.title,'success');
    }
}

export const refresNote = (id, note )=>({
    type:types.notesUpdate,
    payload:{
        id,
        note:{
            id,
            ...note
        }
    }
})


export const startUploading = (file )=>{
    return async (dispatch,getState)=>{
        const { active:activeNote} = getState().notes;

        Swal.fire({
            title:'Uploading...',
            text:'Plase wait...',
            allowOutsideClick:false,
            onBeforeOpen: () =>{
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        //console.log(fileUrl);
        activeNote.url = fileUrl; //Actualizo el URL QUE TENGO EL STORE previamente lo obtuve con getState
        dispatch(startSaveNote(activeNote));

        Swal.close();//Cierra el loading
    }
}

export const startDeleting = (idNote) =>{
    return  async(dispatch,getState )=>{
        const uid = getState().auth.uid;
        await db.doc(`${uid}/journal/notes/${idNote}`).delete();
        

        //Actualizo mi store solo ese compnente
        dispatch(deleteNote(idNote));
    }
}


export const deleteNote = (idNote)=>({
    type:types.notesDelete,
    payload:idNote
});


export const noteLogout = ()=>({
    type:types.notesLogoutCleaning
})