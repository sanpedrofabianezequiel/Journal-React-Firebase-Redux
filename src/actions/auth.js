import Swal from 'sweetalert2';


  import { firebase, googleAuthProvider } from '../firabase/firebase-config';
  
  import {types} from '../types/types';
import { noteLogout } from './notes';
  import { finishLoading, startLoading } from './ui';

  //Middlleware
  export const starLoginEmailPasswordMiddleware= (email,password) =>{
    return (dispatch) =>{

        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user}) =>{
                dispatch(loginFunction(user.uid,user.displayName));
                dispatch(finishLoading())
            })
            .catch(e=>{
                //message viene de firebase
                console.log(e);
                Swal.fire('Error',e.message,'error');
                dispatch(finishLoading())
            })
    }
  }



  export const startRegisterWithEmailPassworName = (email, password,name) =>{
      return (dispatch) =>{
        dispatch(startLoading());
          firebase.auth().createUserWithEmailAndPassword(email,password)
            .then( async ({user}) =>{
                //console.log(user);
                await user.updateProfile({displayName:name});

                dispatch(loginFunction(user.uid,user.displayName));
            })
            .catch(e=>{
                dispatch(finishLoading())
                 //message viene de firebase
                 //console.log(e);
                 Swal.fire('Error',e.message,'error');
                 dispatch(finishLoading())
            })
      }
  }

  export const startGoogleLogin= ()=>{
      return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(userCred=>{
               // console.log(userCred)
               const {user} = userCred;
               dispatch(loginFunction(user.uid,user.displayName))
            })
            .catch(e=>{
                 //message viene de firebase
                 //console.log(e);
                 Swal.fire('Error',e.message,'error');
                 dispatch(finishLoading())
            })
      }
  }

  export const loginFunction = ( uid , displayName) =>({
      type:types.login,
      payload:{
          uid,
          displayName
      }
  })


  export const startLogoutMiddelware = ()=>{
      return async ( dispatch ) =>{
          await firebase.auth().signOut();
          dispatch(logout());
          //Limpiamos el NOTES and ACTIVE
          dispatch(noteLogout());
      }
  }

  export const logout=()=>({
      type:types.logout
  })