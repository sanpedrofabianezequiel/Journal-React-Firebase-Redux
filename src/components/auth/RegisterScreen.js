import React from 'react'
import {Link} from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { useDispatch,useSelector } from 'react-redux';
import validator from 'validator';
import { setErrorAction, removeErrorAction } from '../../actions/ui';
import { startRegisterWithEmailPassworName } from '../../actions/auth';


export const RegisterScreen = () => {

    const dispath = useDispatch();
    const {ui:{msg}}= useSelector(state => state); //CallBack
    console.log(msg);
    const [formValues,handleInputeChange,reset] = useForm({
        name:'ezequiel',
        email:'ezequiel@gmail.com',
        password:'123456',
        password2:'123456'
    });

    const {name,email,password,password2} = formValues;

    const handleRegister= (e)=>{
        e.preventDefault();
        //console.log(name,email,password,password2);
        if(isFormValid()){
            dispath(startRegisterWithEmailPassworName(email,password,name));
        }
    }

    const isFormValid= ()=>{
        if(name === ''){
            dispath(setErrorAction('Name is required'));
            return false;
        }else if(!validator.isEmail(email)){
            dispath(setErrorAction('Email is invalid'));
            return false
        }else if( password !== password2 || password.length < 5){
            dispath(setErrorAction('Password is not valid'));
            return false;
        }
        dispath(removeErrorAction());
        return true;
    }

    return (
        <>
        <h3 className="auth__title">Register</h3>

        {
            msg && (
                <div className="auth__alert-error">
                    {msg}
                </div>
            )
        }

        <form onSubmit={handleRegister}>
            
            <input  value={name} onChange={handleInputeChange}  type="text" placeholder="Name" name="name"  className="auth__input" autoComplete="off" />

            <input  value={email} onChange={handleInputeChange} type="text" placeholder="email" name="email"  className="auth__input" autoComplete="off" />
            
            <input  value={password} onChange={handleInputeChange} type="password" placeholder="Password" name="password" className="auth__input" />

            <input  value={password2} onChange={handleInputeChange} type="password" placeholder="Confirm Password" name="password2" className="auth__input" />

            <button type="submit" className="btn btn-primary btn-block  mb-5 "  >
                Login
            </button>
            <hr/>

           

            <Link to="/auth/login" className="link  ">
                Already registered? 
            </Link>
        </form>
        </>
    )
}
