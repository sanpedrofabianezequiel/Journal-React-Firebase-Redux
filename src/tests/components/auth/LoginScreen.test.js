import React from 'react';
import { mount,shallow} from 'enzyme';
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
const middlewares = [thunk];
const mockStore = configureStore(middlewares)

const initState  = {
    auth:{},
    ui:{
        loading:false,
        msgError:null
    }
};

let store =  mockStore(initState);
store.dispatch= jest.fn();

import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startGoogleLogin,starLoginEmailPasswordMiddleware } from '../../../actions/auth';
jest.mock('../../../components/auth/LoginScreen',()=>({
    startGoogleLogin: jest.fn(),
    starLoginEmailPasswordMiddleware:jest.fn()
}));



const wrapper = shallow(
    <Provider store = {store}>
        <MemoryRouter>
            <LoginScreen/>
        </MemoryRouter>
    </Provider>
);


describe('Pruebas en <LoginScreen />', () => {    

    beforeEach(()=>{
        store =  mockStore(initState);
        jest.clearAllMocks();
    });

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    }); 

    //MOUNT VERSION INCOPATIBLE
    /*test('Debe de disparar la accion de startGoogleLogin', () => {
        wrapper.find('.google-btn').prop('onClick')();
        expect( startGoogleLogin).toHaveBeenCalled();
    })
    
    test('Debe de disparar la accion starLoginEmailPasswordMiddleware',()=>{
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        expect( starLoginEmailPasswordMiddleware).toHaveBeenLastCalledWith('noelia@gmail.com','123456');
    })*/
});
