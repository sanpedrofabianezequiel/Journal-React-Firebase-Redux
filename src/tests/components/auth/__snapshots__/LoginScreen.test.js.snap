// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Debe de mostrarse correctamente 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <MemoryRouter>
    <Component />
  </MemoryRouter>
</ContextProvider>
`;
